spicy/runtime/src/util.cc

```diff
@@ -20,11 +20,44 @@ std::string spicy::rt::version() {
 #endif
 }
 
+
+static inline void byte_to_hex(unsigned char byte, char* hex_out) {
+    static constexpr char hex_chars[] = "0123456789ABCDEF";
+    hex_out[0] = hex_chars[(byte & 0xf0) >> 4];
+    hex_out[1] = hex_chars[byte & 0x0f];
+}
+
 std::string spicy::rt::bytes_to_hexstring(const hilti::rt::Bytes& value) {
+    const auto& data = value.str();
+
+    if ( data.empty() )
+        return "";
+
     std::string result;
+    result.resize(data.size() * 2); // 2 digits per hex byte
+
+    for ( unsigned long i = 0; i < data.size(); i++ )
+        byte_to_hex(data[i], &result[i * 2]);
 
-    for ( auto x : value )
-        result += hilti::rt::fmt("%02X", x);
+    return result;
+}
+
+std::string spicy::rt::bytes_to_mac(const hilti::rt::Bytes& value) {
+    const auto& data = value.str();
+
+    if ( data.empty() )
+        return "";
+
+    std::string result;
+    result.resize((data.size() * 2) +
+                  (data.size() - 1)); // 2 digits per hex byte, plus one colon per byte except the last
+
+    for ( unsigned long i = 0; i < data.size(); i++ ) {
+        if ( i > 0 )
+            result[(i * 3) - 1] = ':';
+
+        byte_to_hex(data[i], &result[i * 3]);
+    }
```

Comment:
Nit: we could allocate `result` with the correct size immediately, and also avoid the branch inserting `:` below if we initialize the full result with `:`s, e.g.,

```suggestion
    // Two digits per hex byte, plus one colon per byte except the last.
    std::string result((data.size() * 2) + (data.size() - 1), ':');

    for ( unsigned long i = 0; i < data.size(); i++ )
        byte_to_hex(data[i], &result[i * 3]);
```

Reply:
Actually I think we want `reserve()`, not `resize()`, so that we don't initialize the data twice. I'll fix that.

Reply:
Since `reserve` only changes the capacity, but not the size that would mean replacing all current element updates via `[..]` with `push_back` which would probably be more expensive.

Reply:
> Actually I think we want `reserve()`, not `resize()`, so that we don't initialize the data twice. I'll fix that.

Never mind, then we can't directly access the indices. I'll take your version then.